### PR TITLE
zsh: update zsh completion for docker command

### PR DESCRIPTION
**Pull request by vincentbernat**
zsh completion is updated with the content of
felixr/docker-zsh-completion.
- felixr/docker-zsh-completion@a93e1cb7bd5a Fix completion of repositories with tags
- felixr/docker-zsh-completion@590ea7059698 Respect provided `--host` flag when invoking docker
- felixr/docker-zsh-completion@6c557babaa3c Several cosmetic improvements
- felixr/docker-zsh-completion@5b63cc591a5c Update completion for `inspect`
- felixr/docker-zsh-completion@b7d8f2f7cc92 Order completions alphabetically
- felixr/docker-zsh-completion@63f6a0622496 Factor completion for `build`, `create` and `run`
- felixr/docker-zsh-completion@ade49ee47f9c Enforce positional arguments being last
- felixr/docker-zsh-completion@850b6b6d95e6 Update completion for build/commit/export/exec/history/import
- felixr/docker-zsh-completion@01bfd8c075dc Remove completion for `insert` and duplicate of `import`
- felixr/docker-zsh-completion@c64a1d730a90 Update completion for `stats` to add `--no-stream` flag
- felixr/docker-zsh-completion@5e81d78b5270 Update completion for `log` to add `--since` flag
- felixr/docker-zsh-completion@b3c146a1a222 Update completion for `run` to add `--group-add` flag
- felixr/docker-zsh-completion@8d4f196ad825 Don't trigger expensive completion function for flags
- felixr/docker-zsh-completion@bd5aaa124d9d Add completion for `--help` everywhere
- felixr/docker-zsh-completion@3a67a0e8c421 Return appropriate status code on completion
- felixr/docker-zsh-completion@4dfcb450eaeb Add Steve as a regular contributor.
- felixr/docker-zsh-completion@996a1c6def35 Add completion for top-level flags
- felixr/docker-zsh-completion@b6df75905f2e Ensure short/long option are not allowed twice
- felixr/docker-zsh-completion@75b6a500a04a Complete repositories with tags only on repository match
- felixr/docker-zsh-completion@5e6292135fcc Factorize completion of images/repositories/tags
- felixr/docker-zsh-completion@1c504eb6779d Handle repositories with ":"
- felixr/docker-zsh-completion@0a05bf818bbc Update completion for `pause' and`unpause'
- felixr/docker-zsh-completion@b3a63253e2bc Containers name can include Swarm host

In summary:
- Swarm support
- Handling repositories with ":"
- Rework how completion of images/repositories/tags work:
  - felixr/docker-zsh-completion@5e6292135fcc
  - felixr/docker-zsh-completion@75b6a500a04a
  - felixr/docker-zsh-completion@a93e1cb7bd5a

The remaining changes are here to sync changes done in Docker repository
(mostly from PR #14074 and #14555, by @sdurrheimer). With some minor changes:
- boolean flags don't complete their arguments (true/false)
- reuse of `--host` argument is done with `$opt_arg` to avoid parsing
  error
- build/create/run common options are factorized out
- `--help` flag is handled differently
- `pause` and `unpause` accepts several containers as far as I know, so
  the change is reverted
- some more, but difficult to notice (more completion for some flags I think)

Some labels are reverted, mostly because I did the merge by copy/pasting
new options instead of modifying existing options.

This commit is partial. The way the `--help` option is handled triggered
a major change due to the way things are quoted. Those changes were
partially and programmaticaly reverted in this commit only to minimize
the changes to review. The next commit will restore the full changes.

Signed-off-by: Vincent Bernat vincent@bernat.im

### CODE REVIEW

**Comment by vincentbernat** [comment]
The missing signature on the second commit is voluntary to avoid accidental merge until review.

**Comment by thaJeztah** [comment]
Thanks for this, @vincentbernat ! Perhaps you can prefix the title of this PR with `[WIP]`, and move the "do not merge" warning to the top of the description in bold. In most cases, PRs with a missing sign-off should not be reviewed ;-)

I'm no expert on (zsh) completions, but could you point out "what" you would like to have reviewed at this moment? (Then I can try and find people for reviewing)

**Comment by vincentbernat** [comment]
I have signed-off the second commit. This is not much a WIP than in need of a review by @sdurrheimer that I didn't undo his changes. Oh, and have one or two other people test wouldn't hurt too. ;-)

File: contrib/completion/zsh/_docker [hunk 1]
```diff
@@ -186,122 +203,126 @@ __docker_commands() {
...
+        "($help)--name=-[Container name]:name: "
+        "($help)--net=-[Network mode]:network mode:(bridge none container host)"
+        "($help)--oom-kill-disable[Disable OOM Killer]"
+        "($help -P --publish-all)"{-P,--publish-all}"[Publish all exposed ports]"
+        "($help)*"{-p,--publish=-}"[Expose a container's port to the host]:port:_ports"
+        "($help)--pid=-[PID namespace to use]:PID: "
+        "($help)--privileged[Give extended privileges to this container]"
+        "($help)--read-only[Mount the container's root filesystem as read only]"
+        "($help)--restart=-[Restart policy]:restart policy:(no on-failure always)"
+        "($help)--rm[Remove intermediate containers when it exits]"
```
Comment (sdurrheimer) [inline contrib/completion/zsh/_docker]: `--rm` only for `docker run`

File: contrib/completion/zsh/_docker [hunk 2]
```diff
@@ -186,122 +203,126 @@ __docker_commands() {
...
 __docker_subcommand() {
-    local -a _command_args
+    local -a _command_args opts_help opts_cpumem opts_create
+    local expl help="-h --help"
+    integer ret=1
+
+    opts_help=("(: -)"{-h,--help}"[Print usage]")
+    opts_cpumem=(
+        "($help -c --cpu-shares)"{-c,--cpu-shares=-}"[CPU shares (relative weight)]:CPU shares:(0 10 100 200 500 800 1000)"
+        "($help)--cpuset=-[CPUs in which to allow execution]:CPU set: "
```
Comment (sdurrheimer) [inline contrib/completion/zsh/_docker]: `--cpuset` is replaced by `--cpuset-cpus`, isn't it ?

File: contrib/completion/zsh/_docker [hunk 3]
```diff
@@ -186,122 +203,126 @@ __docker_commands() {
...
+        "($help -a --attach)"{-a,--attach=-}"[Attach to stdin, stdout or stderr]:device:(STDIN STDOUT STDERR)"
+        "($help)*--add-host=-[Add a custom host-to-IP mapping]:host\:ip mapping: "
+        "($help)--blkio-weight=-[Block IO (relative weight), between 10 and 1000]:Block IO weight:(10 100 500 1000)"
+        "($help)*--cap-add=-[Add Linux capabilities]:capability: "
+        "($help)*--cap-drop=-[Drop Linux capabilities]:capability: "
+        "($help)--cidfile=-[Write the container ID to the file]:CID file:_files"
+        "($help)*--device=-[Add a host device to the container]:device:_files"
+        "($help)*--dns=-[Set custom dns servers]:dns server: "
+        "($help)*--dns-search=-[Set custom DNS search domains]:dns domains: "
+        "($help)*"{-e,--environment=-}"[Set environment variables]:environment variable: "
```
Comment (sdurrheimer) [inline contrib/completion/zsh/_docker]: Should be `--env` instead of `--environment`

File: contrib/completion/zsh/_docker [hunk 4]
```diff
@@ -309,239 +330,189 @@ __docker_subcommand() {
...
             _arguments \
-                '(-f --follow)'{-f,--follow}'[Follow log output]' \
-                '(- :)--help[Print usage]' \
-                '--since=-[Show logs since timestamp]:timestamp: ' \
-                '(-t --timestamps)'{-t,--timestamps}'[Show timestamps]' \
-                '--tail=-[Number of lines to show from the end of the logs]:lines:(1 10 20 50 all)' \
-                '*:containers:__docker_containers' && ret=0
+                $opts_help \
+                "($help -f --follow)"{-f,--follow}"[Follow log output]" \
+                "($help -t --timestamps)"{-t,--timestamps}"[Show timestamps]" \
```
Comment (sdurrheimer) [inline contrib/completion/zsh/_docker]: nit: `--timestamps` after `--since`

**Comment by sdurrheimer** [comment]
I have made some comments.

Testing the whole completion script at the same time.

**Comment by vincentbernat** [comment]
Thanks! I have pushed an additional commit to fix the problems you spotted. If all is OK with you, I'll squash the three commits.

File: contrib/completion/zsh/_docker [hunk 5]
```diff
@@ -615,70 +584,59 @@ _docker() {
...
+        "($help)--default-gateway-v6[Container default gateway IPv6 address]:IPv6 address: " \
+        "($help)*--dns=-[DNS server to use]:DNS: " \
+        "($help)*--dns-search=-[DNS search domains to use]" \
+        "($help)*--default-ulimit=-[Set default ulimit settings for containers]:ulimit: " \
+        "($help -e --exec-driver)"{-e,--exec-driver=-}"[Exec driver to use]:driver:(native lxc windows)" \
+        "($help)*--exec-opt=-[Set exec driver options]:exec driver options: " \
+        "($help)--exec-root=-[Root of the Docker execdriver (default: /var/run/docker)]:path:_directories" \
+        "($help)--fixed-cidr=-[IPv4 subnet for fixed IPs]:IPv4 subnet: " \
+        "($help)--fixed-cidr-v6=-[IPv6 subnet for fixed IPs]:IPv6 subnet: " \
+        "($help  -G --group)"{-G,--group=-}"[Group for the unix socket (default: docker)]:group:_groups" \
```
Comment (sdurrheimer) [inline contrib/completion/zsh/_docker]: nit: double space to single space. Further below.

**Comment by sdurrheimer** [comment]
One last comment, otherwise LGTM :+1:

**Comment by vincentbernat** [comment]
Updated and all commits have been squashed into one.

**Comment by thaJeztah** [comment]
ping @albers and @tianon I know zsh is not your thing, but perhaps you could have a look so that this can be moved forward? Thanks!

**Comment by jfrazelle** [comment]
LGTM

**Comment by jfrazelle** [comment]
thanks so much!

**Comment by thaJeztah** [comment]
thanks both, for contributing and reviewing! :+1:

**Comment by albers** [comment]
@thaJeztah I'm not yet familiar enough with zsh and its completion. In fact I do not even have it installed yet, so I could not verify the PR quickly.
I promise to look into zsh completion soon.

**Comment by thaJeztah** [comment]
@albers that would be great (I'm not familiar with completion *at all*, so :smile:)

You could even consider running it in a container, similar to: https://registry.hub.docker.com/u/jess/fish/

**Comment by jfrazelle** [comment]
yes omg i was going to link that \o/

**Comment by thaJeztah** [comment]
@jfrazelle don't think you have a zsh image (yet)?

**Comment by jfrazelle** [comment]
I think there is a PR for one

**Pull request merged**